docs/Dialects/Interop/RationaleInterop.md

```diff
@@ -0,0 +1,590 @@
+and everything surrounding it using the CIRCT-based simulator? This dialect can
+provide the latter by implementing an interop operation and one lowering per
+simulator rather than custom pair-wise interop implementations.
+
+We make a clear distinction between procedural and structural interop because
+it is not possible to provide direct interop between a purely structural
+language and a purely procedural language. To achieve interop between them,
+one has to use at least one bridging language that has (some) support for both
+as a translation layer.
+
+## Procedural Interop
+
+The `interop` dialect provides a few operations to represent partially lowered
+procedural interop instances presented in the
+[next chapter](#representing-partial-lowerings) as well as an interface to be
+implemented by the surrounding module and rewrite patterns to call into those
+interface implementations presented in the chapter about
+['Container-side Lowering'](#container-side-lowering).
+It also provides common rewrite patterns to insert translation layers between
+different interop mechanisms like textural C++, C-foreign-functions, Verilog,
+etc. They are discussed in more detail in the chapters
+['Interop Mechanisms'](#interop-mechanisms) and
+['Bridging between Interop Mechanisms'](#bridging-between-interop-mechanisms).
+
+### Representing Partial Lowerings
+
+There are four operations to represent partially lowered procedural interop in
+the IR:
+* `interop.procedural.alloc`
+* `interop.procedural.init`
+* `interop.procedural.update`
+* `interop.procedural.dealloc`
+
+The `alloc` operation returns a variadic list of values that represent
+persistent state, i.e., state that has to persist across multiple executions
+of the `update` operation. For example, it can be lowered to C++ class fields
+that are persistent across multiple calls of a member function, or to global
+simulator state that persists over simulation cycles, etc.
+
+Additionally, it has an attribute that specifies the interop mechanism under
+which the state types are valid. This is necessary to allow bridging patterns
+to map the types to valid types in the other interop mechanism, e.g., to an
+opaque pointer, if it does not support the same types.
+
+```mlir
+%state = interop.procedural.alloc cpp : !emitc.ptr<!emitc.opaque<"VBar">>
+```
+
+The `init` operation takes the variadic list of states from the `alloc`
+operation as operands and has a body with a `return` operation that has a
+variadic list of operands that matches the types of the states and
+represent the initial values to be assigned to the state values. The assignment
+will be inserted by the container-side lowering of the interop operations.
+The operation also has an interop mechanism attribute for the same reason as
+above and, additionally, to wrap the operations in the body in a way to make
+them executable in the other interop mechanism, e.g., wrap them in a
+`extern "C"` function to make it callable from C or LLVM IR.
```

Comment:
It would be good to identify the `cpp` bit as the "interop mechanism attribute".